Implemented option :include_captures in Regex.split/3

druzn3k opened the pull request:
Closes issue #4203. Not the prettiest implementation, but it does work. Any suggestion on how to make it better, please?

lexmag commented on lib/elixir/lib/regex.ex:
```diff
@@ -348,6 +348,9 @@ defmodule Regex do
       order. Defaults to `:first` which means captures inside the regex do not
       affect the splitting process.
 
+    * `:include_captures` - when `true`, includes in the result the matches of
```
It feels `:include_` prefix might be redundant.

druzn3k replied:
dunno, using only `:captures` feels not clear enough to me, if I have to think about reading code wrote by someone else. es: `Regex.split(~r/abc/, "abcdefg", captures: true)`. Who captures what? I think that removing the prefix is not the right choice.

lexmag replied:
Capture is very natural name in regex handling, also we already have :capture option in `run` function.
Maybe `with_` prefix since we use it for variable name?

josevalim replied:
`run` uses capture as a verb. I am `include_captures` works better than `with_captures` for me (because of the verb).

RobertDober replied:
I really like `include_captures` :+1:

josevalim commented on lib/elixir/lib/regex.ex:
```diff
@@ -397,30 +404,40 @@ defmodule Regex do
+    do: do_split(t, string, offset, counter, trim, with_captures)
 
-  defp do_split([[{pos, length}|h]|t], string, offset, counter, trim) do
+  defp do_split([[{pos, length}|h]|t], string, offset, counter, trim, with_captures) do
```
I am wondering if we should rather match on `with_captures` here are the head and have two clauses, one for when it is true and another for when it is false?

druzn3k replied:
I thought about that, but the differences between the two functions seems too small to be that redundant. Another choice could be matching the truth table of `{trim, with_captures}`. Should I?

josevalim replied:
I think the logic today, specially around the first if, is a bit convoluted. We also have many nested IFs, so we definitely need to refactor. Moving the apart may be the best way to find out how they are similar and move on from there.

druzn3k replied:
I'm splitting the function. Keep you posted.

josevalim commented on lib/elixir/lib/regex.ex:
```diff
@@ -397,30 +404,40 @@ defmodule Regex do
-      <<_::binary-size(offset), part::binary-size(keep), _::binary>> = string
-      [part|do_split([h|t], string, new_offset, counter - 1, trim)]
+      if with_captures do
+        <<_::binary-size(offset), part::binary-size(keep), mtch::binary-size(length), _::binary>> = string
```
Let's call it `match` instead of `mtch`.

druzn3k replied:
Sure, refactoring that! :+1:

lexmag commented on lib/elixir/lib/regex.ex:
```diff
@@ -397,30 +404,40 @@ defmodule Regex do
+        if keep == 0 and (length == 0 or trim) do
+          [mtch|do_split([h|t], string, new_offset, counter - 1, trim, with_captures)]
+        else
+          [part|[mtch|do_split([h|t], string, new_offset, counter - 1, trim, with_captures)]]
```
Please put spaces around `|` operator.

druzn3k replied:
Ok!

druzn3k commented:
@josevalim @lexmag new version with the requested changes online!

lexmag commented on lib/elixir/lib/regex.ex:
```diff
@@ -397,30 +404,72 @@ defmodule Regex do
+    end
+  end
+
+  defp do_split([[{pos, length}|h]|t], string, offset, counter, true, true) do
```
I think previous version (single clause) was easier to comprehend and maintain.

lexmag replied:
Looks like wrapping `trim` and `with_captures` options in a tuple will reduce the visual noise and chance of error.

druzn3k replied:
Should I revert the last changes?

lexmag replied:
I'd say yes, @josevalim what are you thoughts?

lexmag replied:
Also `cond` usage makes single clause definition much better. :bowtie:

druzn3k replied:
:+1:

josevalim replied:
I had originally thought to split only in terms of include_captures, so it would be in 2 clauses and not 4. :) This way only the first 2 lines of each clause would be the same (he ones where we calculate both new_offset and keep). Everything else would be different.

lexmag commented on lib/elixir/lib/regex.ex:
```diff
@@ -397,30 +404,72 @@ defmodule Regex do
+      if keep == 0 do
+        [match | do_split([h|t], string, new_offset, counter - 1, true, true)]
+      else
+        [part | [match | do_split([h|t], string, new_offset, counter - 1, true, true)]]
```
We could simplify it with: `[part, match | do_split...]`

druzn3k replied:
:+1:

lexmag commented on lib/elixir/lib/regex.ex:
```diff
@@ -397,30 +404,72 @@ defmodule Regex do
+    if keep == 0 and length == 0 do
+      do_split([h|t], string, new_offset, counter, true, true)
+    else
+      <<_::binary-size(offset), part::binary-size(keep), match::binary-size(length), _::binary>> = string
```
Common style is to have spaces around `::`.

druzn3k commented:
@josevalim @lexmag changes done! :+1: 

The only change I didn't do is spaces around `::`. I looked around in `regex.ex` and `::` is always without surrounding spaces, so I didn't want to introduce a change without changing the rest of the file. I could do it in another PR, but this PR doesn't seem the appropriate place to me, given it's scope. But if you still want the change I will be happy to do it! :smile:

RobertDober commented:
I really appreciate the acceptance and prompt implementation of this issue. :) 
Now I might seem ungrateful, but please can any good soul explain to me why this code does not simply delegate to `:re.split`.

druzn3k commented:
No worries @RobertDober, not coming out as ungrateful. We are here to make this project better! For me the answer is really simple; I didn't know the existence of `:re.split`! I will explore the possibility to use `:re.split` instead of this, but at a first glance seems to me that behaviour is different, so I think it's not a trivial change. The [whole thing](http://www.erlang.org/doc/man/re.html#split-3) of `/(re)/`vs `/re/` could break existing applications.

RobertDober commented:
@druzn3k BTW `:re.split` implies of course `include_captures: true` so code simplification is probably not possible easily. But I agree on the spirit, and curiosity is just one of my many vices.

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

druzn3k commented:
@RobertDober, no, it doesn't imply `include_captures: true`, it depends on **how** one wrote the regex. If you use a subexpression like `([abc])`it includes the result, otherwise not. This is the relevant part in the documentation:

> If subexpressions are given in the regular expression, the matching subexpressions are returned in the resulting list as well.

This will break half of the tests on the `Regex` alone and a lot of dependent modules. If someone wrote the regex using `(_)` the behaviour will change, breaking the application in a subtle way (wrong data on output), silently failing. It could potentially be a disaster.

druzn3k commented:
Thanks @josevalim !

josevalim commented:
@RobertDober implementing our own allows us to provide more features and performance optimizations in the short term. For example, like the `:on` feature that allows you to specify the capture to split on.

@druzn3k i believe he suggested using `:re.split` only for the branch that you added but again, it wouldn't work with the `:on` option. :)

druzn3k commented:
@josevalim @RobertDober sorry I interpreted it in a "global" way. :+1:

RobertDober commented:
@josevalim that is correct José :). @druzn3k No problem, I admire your energy wrapping your mind around `do_split` and friends. But for backward compatibility I will need to use `:re.split` in Earmark and that was also why I was asking about it just in case there was a problem with it.

Anyway thank you all for your work and the information.

lexmag commented:
@druzn3k regarding style changes, I believe correction of style while you on some code is the way to fix it (regardless of the rest of the file), especially since we don't accept purely style changing PRs.

druzn3k commented:
@RobertDober: I love to contribute anytime I get a chance. :smile: 
@lexmag: sorry, didn't know the last bit (that cosmetic PRs are not accepted). My bad!
